landoapi/hg.py

```diff
@@ -368,56 +372,158 @@ def apply_patch(self, patch_io_buf):
+        Changes made by code formatters are applied to the working directory and
+        are not committed into version control.
+        """
+        return self.run_mach_command(["lint", "--fix", "--outgoing"])
 
-        # Exit if we didn't find any patterns.
-        if not fix_hg_command:
-            return None
+    def run_mach_bootstrap(self) -> str:
+        """Run `mach bootstrap` to configure the system for code formatting."""
+        return self.run_mach_command(
+            [
+                "bootstrap",
+                "--no-system-changes",
+                "--application-choice",
+                "browser",
+            ]
+        )
 
-        # Run the formatters.
-        fix_hg_command += ["fix", "-r", "stack()"]
-        fix_output = self.run_hg(fix_hg_command).splitlines()
+    def run_mach_command(self, args: List[str]) -> str:
+        """Run a command using the local `mach`, raising if it is missing."""
+        if not self.mach_path:
+            raise Exception("No `mach` found in local repo!")
+
+        # Convert to `str` here so we can log the mach path.
+        command_args = [str(self.mach_path)] + args
+
+        try:
+            logger.info("running mach command", extra={"command": command_args})
+
+            output = subprocess.run(
+                command_args,
+                capture_output=True,
+                check=True,
+                cwd=self.path,
+                encoding="utf-8",
+                universal_newlines=True,
+            )
+
+            logger.info(
+                "output from mach command",
+                extra={
+                    "output": output.stdout,
+                },
+            )
+
+            return output.stdout
+
+        except subprocess.CalledProcessError as exc:
+            logger.exception(
+                "Failed to run mach command",
+                extra={
+                    "command": command_args,
+                    "err": exc.stderr,
+                    "output": exc.stdout,
+                },
+            )
 
-        # Update the working directory to the latest change.
-        self.run_hg(["update", "-C", "-r", "tip"])
+            raise exc
 
-        # Exit if no revisions were reformatted.
-        pre_formatting_hashes = check_fix_output_for_replacements(fix_output)
-        if not pre_formatting_hashes:
+    def format_stack_amend(self) -> Optional[List[str]]:
+        """Amend the top commit in the patch stack with changes from formatting."""
+        try:
+            # Amend the current commit, using `--no-edit` to keep the existing commit message.
+            self.run_hg(["commit", "--amend", "--no-edit", "--landing_system", "lando"])
+
+            return [self.get_current_node().decode("utf-8")]
+        except hglib.error.CommandError as exc:
+            if exc.out.strip() == b"nothing changed":
+                # If nothing changed after formatting we can just return.
+                return None
```

Comment:
nit
```suggestion
                return
```

Reply:
Meh, personally I prefer to be explicit that `None` is going to be returned from a function, especially when the type is an `Optional`. I usually only ever do a generic `return` when the function itself would never return something. I'll change it in this case. :+1: